src/Jellyfin.Database/Jellyfin.Database.Implementations/Locking/PessimisticLockBehavior.cs

```diff
@@ -0,0 +1,107 @@
+using System;
+using System.Data.Common;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Diagnostics;
+
+namespace Jellyfin.Database.Implementations.Locking;
+
+/// <summary>
+/// A locking behavior that will always block any operation while a write is requested. Mimikes the old SqliteRepository behavior.
+/// </summary>
+public class PessimisticLockBehavior : IEntityFrameworkCoreLockingBehavior
+{
+    private static ReaderWriterLockSlim DatabaseLock { get; } = new();
+
+    /// <inheritdoc/>
+    public void OnSaveChanges(JellyfinDbContext context, Action saveChanges)
+    {
+        try
+        {
+            DatabaseLock.EnterWriteLock();
+            saveChanges();
+        }
+        finally
+        {
+            DatabaseLock.ExitWriteLock();
+        }
+    }
+
+    /// <inheritdoc/>
+    public void Initialise(DbContextOptionsBuilder optionsBuilder)
+    {
+        optionsBuilder.AddInterceptors(new LockingInterceptor());
+    }
+
+    /// <inheritdoc/>
+    public async Task OnSaveChangesAsync(JellyfinDbContext context, Func<Task> saveChanges)
+    {
+        try
+        {
+            DatabaseLock.EnterWriteLock();
+            await saveChanges().ConfigureAwait(false);
+        }
+        finally
+        {
+            DatabaseLock.ExitWriteLock();
+        }
+    }
+
+    /// <summary>
+    /// Adds strict read/write locking.
+    /// </summary>
+    private sealed class LockingInterceptor : DbCommandInterceptor
+    {
+        /// <inheritdoc/>
+        public override InterceptionResult<DbDataReader> ReaderExecuting(
+            DbCommand command,
+            CommandEventData eventData,
+            InterceptionResult<DbDataReader> result)
+        {
+#pragma warning disable MT1012 // Acquiring lock without guarantee of releasing
+            if (!DatabaseLock.IsWriteLockHeld) // enter a read lock only if not already a write lock has been issued by the savechanges invocation
+            {
+                DatabaseLock.EnterReadLock();
```

Comment:
Doesn't the read lock here prevent other reads from happening at the same time?

Reply:
No read locks do not block read locks

> [ReaderWriterLock](https://learn.microsoft.com/en-us/dotnet/api/system.threading.readerwriterlock?view=net-9.0) is used to synchronize access to a resource. At any given time, it allows either concurrent read access for multiple threads, or write access for a single thread.